File: sdk/typescript/src/services/policy.ts
/**
 * ACGS-2 Policy Service
 * Constitutional Hash: cdd01ef066bc6cf2
 */

import { ACGS2Client } from '../client';
import {
  CONSTITUTIONAL_HASH,
  Policy,
  PolicySchema,
  PolicyStatus,
  CreatePolicyRequest,
  UpdatePolicyRequest,
  PaginationParams,
  PaginatedResponse,
  ApiResponse,
} from '../types';
import { ValidationError } from '../utils';

// =============================================================================
// Policy Service
// =============================================================================

export class PolicyService {
  private readonly basePath = '/api/v1/policies';

  constructor(private readonly client: ACGS2Client) {}

  // ===========================================================================
  // CRUD Operations
  // ===========================================================================

  /**
   * Creates a new policy
   */
  async create(request: CreatePolicyRequest): Promise<Policy> {
    const response = await this.client.post<Policy>(this.basePath, {
      ...request,
      constitutionalHash: CONSTITUTIONAL_HASH,
    });

    if (!response.success || !response.data) {
      throw new ValidationError('Failed to create policy');
    }

    return this.validatePolicy(response.data);
  }

  /**
   * Retrieves a policy by ID
   */
  async get(policyId: string): Promise<Policy> {
    const response = await this.client.get<Policy>(`${this.basePath}/${policyId}`);

    if (!response.success || !response.data) {
      throw new ValidationError(`Policy not found: ${policyId}`);
    }

    return this.validatePolicy(response.data);
  }

  /**
   * Updates a policy
   */
  async update(policyId: string, request: UpdatePolicyRequest): Promise<Policy> {
    const response = await this.client.patch<Policy>(`${this.basePath}/${policyId}`, request);

    if (!response.success || !response.data) {
      throw new ValidationError(`Failed to update policy: ${policyId}`);
    }

    return this.validatePolicy(response.data);
  }

  /**
   * Deletes a policy
   */
  async delete(policyId: string): Promise<void> {
    await this.client.delete(`${this.basePath}/${policyId}`);
  }

  /**
   * Lists policies with pagination
   */
  async list(params?: PaginationParams & {
    status?: PolicyStatus;
    tags?: string[];
  }): Promise<PaginatedResponse<Policy>> {
    const response = await this.client.get<PaginatedResponse<Policy>>(this.basePath, params);

    if (!response.success || !response.data) {
      throw new ValidationError('Failed to list policies');
    }

    // Validate each policy
    response.data.data = response.data.data.map((policy) => this.validatePolicy(policy));
    return response.data;
  }

  // ===========================================================================
  // Status Management
  // ===========================================================================

  /**
   * Activates a policy
   */
  async activate(policyId: string): Promise<Policy> {
    return this.updateStatus(policyId, PolicyStatus.ACTIVE);
  }

  /**
   * Deprecates a policy
   */
  async deprecate(policyId: string): Promise<Policy> {
    return this.updateStatus(policyId, PolicyStatus.DEPRECATED);
  }

  /**
   * Archives a policy
   */
  async archive(policyId: string): Promise<Policy> {
    return this.updateStatus(policyId, PolicyStatus.ARCHIVED);
  }

  /**
   * Submits a policy for review
   */
  async submitForReview(policyId: string): Promise<Policy> {
    return this.updateStatus(policyId, PolicyStatus.PENDING_REVIEW);
  }

  /**
   * Approves a policy
   */
  async approve(policyId: string): Promise<Policy> {
    return this.updateStatus(policyId, PolicyStatus.APPROVED);
  }

  private async updateStatus(policyId: string, status: PolicyStatus): Promise<Policy> {
    const response = await this.client.post<Policy>(
      `${this.basePath}/${policyId}/status`,
      { status }
    );

    if (!response.success || !response.data) {
      throw new ValidationError(`Failed to update policy status: ${policyId}`);
    }

    return this.validatePolicy(response.data);
  }

  // ===========================================================================
  // Versioning
  // ===========================================================================

  /**
   * Gets all versions of a policy
   */
  async getVersions(policyId: string): Promise<Policy[]> {
    const response = await this.client.get<Policy[]>(`${this.basePath}/${policyId}/versions`);

    if (!response.success || !response.data) {
      throw new ValidationError(`Failed to get policy versions: ${policyId}`);
    }

    return response.data.map((policy) => this.validatePolicy(policy));
  }

  /**
   * Gets a specific version of a policy
   */
  async getVersion(policyId: string, version: string): Promise<Policy> {
    const response = await this.client.get<Policy>(
      `${this.basePath}/${policyId}/versions/${version}`
    );

    if (!response.success || !response.data) {
      throw new ValidationError(`Policy version not found: ${policyId}@${version}`);
    }

    return this.validatePolicy(response.data);
  }

  /**
   * Creates a new version of a policy
   */
  async createVersion(policyId: string, changes: UpdatePolicyRequest): Promise<Policy> {
    const response = await this.client.post<Policy>(
      `${this.basePath}/${policyId}/versions`,
      changes
    );

    if (!response.success || !response.data) {
      throw new ValidationError(`Failed to create policy version: ${policyId}`);
    }

    return this.validatePolicy(response.data);
  }

  // ===========================================================================
  // Validation & Analysis
  // ===========================================================================

  /**
   * Validates policy rules syntax
   */
  async validate(rules: Record<string, unknown>[]): Promise<{
    valid: boolean;
    errors: string[];
  }> {
    const response = await this.client.post<{ valid: boolean; errors: string[] }>(
      `${this.basePath}/validate`,
      { rules, constitutionalHash: CONSTITUTIONAL_HASH }
    );

    if (!response.success || !response.data) {
      throw new ValidationError('Failed to validate policy rules');
    }

    return response.data;
  }

  /**
   * Analyzes policy impact
   */
  async analyzeImpact(policyId: string): Promise<{
    affectedAgents: string[];
    riskLevel: 'low' | 'medium' | 'high';
    estimatedImpact: string;
    recommendations: string[];
  }> {
    const response = await this.client.get<{
      affectedAgents: string[];
      riskLevel: 'low' | 'medium' | 'high';
      estimatedImpact: string;
      recommendations: string[];
    }>(`${this.basePath}/${policyId}/impact`);

    if (!response.success || !response.data) {
      throw new ValidationError(`Failed to analyze policy impact: ${policyId}`);
    }

    return response.data;
  }

  /**
   * Compares two policy versions
   */
  async diff(policyId: string, fromVersion: string, toVersion: string): Promise<{
    additions: Record<string, unknown>[];
    deletions: Record<string, unknown>[];
    modifications: Array<{
      path: string;
      from: unknown;
      to: unknown;
    }>;
  }> {
    const response = await this.client.get<{
      additions: Record<string, unknown>[];
      deletions: Record<string, unknown>[];
      modifications: Array<{ path: string; from: unknown; to: unknown }>;
    }>(`${this.basePath}/${policyId}/diff`, { fromVersion, toVersion });

    if (!response.success || !response.data) {
      throw new ValidationError(`Failed to diff policy versions: ${policyId}`);
    }

    return response.data;
  }

  // ===========================================================================
  // Helpers
  // ===========================================================================

  private validatePolicy(data: unknown): Policy {
    const result = PolicySchema.safeParse(data);
    if (!result.success) {
      throw new ValidationError('Invalid policy data', {
        validation: result.error.errors.map((e) => e.message),
      });
    }
    return result.data;
  }
}

File: src/core/sdk/python/acgs2_sdk/services/policy.py
"""
ACGS-2 Policy Service
Constitutional Hash: cdd01ef066bc6cf2
"""

from typing import TYPE_CHECKING, Any, Dict, List, Union
try:
    from src.core.shared.types import JSONDict, JSONValue
except ImportError:
    JSONValue = Union[str, int, float, bool, None, Dict[str, Any], List[Any]]  # type: ignore[misc]
    JSONDict = Dict[str, JSONValue]  # type: ignore[misc]

from acgs2_sdk.constants import CONSTITUTIONAL_HASH, POLICIES_ENDPOINT
from acgs2_sdk.models import (
    CreatePolicyRequest,
    PaginatedResponse,
    Policy,
    PolicyStatus,
    UpdatePolicyRequest,
)

if TYPE_CHECKING:
    from acgs2_sdk.client import ACGS2Client


class PolicyService:
    """Service for managing policies."""

    def __init__(self, client: "ACGS2Client") -> None:
        """Initialize the policy service.

        Args:
            client: ACGS2 client instance
        """
        self._client = client
        self._base_path = POLICIES_ENDPOINT

    async def create(self, request: CreatePolicyRequest) -> Policy:
        """Create a new policy.

        Args:
            request: Policy creation request

        Returns:
            Created policy
        """
        data = await self._client.post(
            self._base_path,
            json={
                **request.model_dump(by_alias=True, exclude_none=True),
                "constitutionalHash": CONSTITUTIONAL_HASH,
            },
        )
        return Policy.model_validate(data.get("data", data))

    async def get(self, policy_id: str) -> Policy:
        """Get a policy by ID.

        Args:
            policy_id: Policy ID

        Returns:
            Policy details
        """
        data = await self._client.get(f"{self._base_path}/{policy_id}")
        return Policy.model_validate(data.get("data", data))

    async def update(self, policy_id: str, request: UpdatePolicyRequest) -> Policy:
        """Update a policy.

        Args:
            policy_id: Policy ID
            request: Policy update request

        Returns:
            Updated policy
        """
        data = await self._client.patch(
            f"{self._base_path}/{policy_id}",
            json=request.model_dump(by_alias=True, exclude_none=True),
        )
        return Policy.model_validate(data.get("data", data))

    async def delete(self, policy_id: str) -> None:
        """Delete a policy.

        Args:
            policy_id: Policy ID
        """
        await self._client.delete(f"{self._base_path}/{policy_id}")

    async def list_policies(
        self,
        page: int = 1,
        page_size: int = 50,
        status: PolicyStatus | None = None,
        tags: list[str] | None = None,
    ) -> PaginatedResponse[Policy]:
        """List policies.

        Args:
            page: Page number
            page_size: Items per page
            status: Filter by status
            tags: Filter by tags

        Returns:
            Paginated list of policies
        """
        params: JSONDict = {"page": page, "pageSize": page_size}
        if status:
            params["status"] = status.value
        if tags:
            params["tags"] = ",".join(tags)

        data = await self._client.get(self._base_path, params=params)
        response_data = data.get("data", data)
        return PaginatedResponse[Policy](
            data=[Policy.model_validate(p) for p in response_data.get("data", [])],
            total=response_data.get("total", 0),
            page=response_data.get("page", page),
            page_size=response_data.get("pageSize", page_size),
            total_pages=response_data.get("totalPages", 0),
        )

    async def activate(self, policy_id: str) -> Policy:
        """Activate a policy.

        Args:
            policy_id: Policy ID

        Returns:
            Activated policy
        """
        data = await self._client.post(
            f"{self._base_path}/{policy_id}/status",
            json={"status": PolicyStatus.ACTIVE.value},
        )
        return Policy.model_validate(data.get("data", data))

    async def deprecate(self, policy_id: str) -> Policy:
        """Deprecate a policy.

        Args:
            policy_id: Policy ID

        Returns:
            Deprecated policy
        """
        data = await self._client.post(
            f"{self._base_path}/{policy_id}/status",
            json={"status": PolicyStatus.DEPRECATED.value},
        )
        return Policy.model_validate(data.get("data", data))

    async def validate_rules(self, rules: list[JSONDict]) -> JSONDict:
        """Validate policy rules syntax.

        Args:
            rules: Policy rules to validate

        Returns:
            Validation result with errors if any
        """
        data = await self._client.post(
            f"{self._base_path}/validate",
            json={"rules": rules, "constitutionalHash": CONSTITUTIONAL_HASH},
        )
        return data.get("data", data)

    async def analyze_impact(self, policy_id: str) -> JSONDict:
        """Analyze policy impact.

        Args:
            policy_id: Policy ID

        Returns:
            Impact analysis result
        """
        data = await self._client.get(f"{self._base_path}/{policy_id}/impact")
        return data.get("data", data)

    async def get_versions(self, policy_id: str) -> list[Policy]:
        """Get all versions of a policy.

        Args:
            policy_id: Policy ID

        Returns:
            List of policy versions
        """
        data = await self._client.get(f"{self._base_path}/{policy_id}/versions")
        return [Policy.model_validate(p) for p in data.get("data", [])]
